src/merkle/tiered_smt/tests.rs

```diff
@@ -252,6 +257,209 @@ fn tsmt_update() {
+    let value_a = [ONE, ONE, ONE, ONE];
+    smt.insert(key_a, value_a);
+
+    // --- insert another with the same 32-bit prefix into the tree -----------
+    let smt1 = smt.clone();
+    let raw_b = 0b_01010101_01010101_11111111_11111111_11010110_10010011_11100000_00000000_u64;
+    let key_b = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_b)]);
+    let value_b = [ONE, ONE, ONE, ZERO];
+    smt.insert(key_b, value_b);
+
+    // --- insert the 3rd value with the same 32-bit prefix into the tree -----
+    let smt2 = smt.clone();
+    let raw_c = 0b_01010101_01010101_11111111_11111111_11110110_10010011_11100000_00000000_u64;
+    let key_c = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_c)]);
+    let value_c = [ONE, ONE, ZERO, ZERO];
+    smt.insert(key_c, value_c);
+
+    // --- delete the last inserted value -------------------------------------
+    assert_eq!(smt.insert(key_c, [ZERO; 4]), value_c);
+    assert_eq!(smt, smt2);
+
+    // --- delete the last inserted value -------------------------------------
+    assert_eq!(smt.insert(key_b, [ZERO; 4]), value_b);
+    assert_eq!(smt, smt1);
+
+    // --- delete the first inserted value ------------------------------------
+    assert_eq!(smt.insert(key_a, [ZERO; 4]), value_a);
+    assert_eq!(smt, smt0);
+}
+
+#[test]
+fn tsmt_delete_48_mixed_prefix() {
+    let mut smt = TieredSmt::default();
+
+    // test the case when some values share a 32-bit prefix and others share a 16-bit prefix
+
+    // --- insert a value into the tree ---------------------------------------
+    let smt0 = smt.clone();
+    let raw_a = 0b_01010101_01010101_11111111_11111111_10010110_10010011_11100000_00000000_u64;
+    let key_a = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_a)]);
+    let value_a = [ONE, ONE, ONE, ONE];
+    smt.insert(key_a, value_a);
+
+    // --- insert another with the same 16-bit prefix into the tree -----------
+    let smt1 = smt.clone();
+    let raw_b = 0b_01010101_01010101_01111111_11111111_10010110_10010011_11100000_00000000_u64;
+    let key_b = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_b)]);
+    let value_b = [ONE, ONE, ONE, ZERO];
+    smt.insert(key_b, value_b);
+
+    // --- insert a value with the same 32-bit prefix as the first value -----
+    let smt2 = smt.clone();
+    let raw_c = 0b_01010101_01010101_11111111_11111111_11010110_10010011_11100000_00000000_u64;
+    let key_c = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_c)]);
+    let value_c = [ONE, ONE, ZERO, ZERO];
+    smt.insert(key_c, value_c);
+
+    // --- insert another value with the same 32-bit prefix as the first value
+    let smt3 = smt.clone();
+    let raw_d = 0b_01010101_01010101_11111111_11111111_11110110_10010011_11100000_00000000_u64;
+    let key_d = RpoDigest::from([ONE, ONE, ONE, Felt::new(raw_d)]);
+    let value_d = [ONE, ZERO, ZERO, ZERO];
+    smt.insert(key_d, value_d);
+
+    // --- delete the inserted values one-by-one ------------------------------
```

Comment:
I don't think that's necessary. Here we keep track of the states of the tree prior to the time when nodes were inserted (`smt0`, `smt1` etc.) and then comparing the tree resulting from a deletion to the state before a given node was inserted. So, if the query returns a wrong value for `smt` it would have returned a wrong value for `smt0`, `smt1` etc. and that would mean that the problem is not with value removal but with something else.